Initial etcd volume recovery support

moelsayed opened the pull request:
**What this PR does / why we need it**:
This PR provides support for automatic recovery in cases where one of the PVs for the user etcd cluster is lost.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5544

**Special notes for your reviewer**:

**Documentation**:
<!-- Add links to the related documentation changes related to this pull request. E.g. the link to the kubermatic/docs pullrequest. -->

**Does this PR introduce a user-facing change?**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If  no release note is required, just write "NONE".
-->
```release-note
NONE
```

moelsayed committed:
Initial etcd volume recovery

multi-io commented:
Tested this on a 3-node cluster, deleted the etcd-1 PV, removing the pv-protection finalizer in the process, also deleted the etcd-1 pod then (not sure if that would've been necessary). Eventually the PV deletion went through, the kubermatic_volume_watcher_controller kicked in, recreated the pvc and pv and restarted all the etcd pods, end result was a working 3-node cluster with all pods restarted and the etcd-1 PVC and PV recreated. 👍

moelsayed commented:
> Tested this on a 3-node cluster, deleted the etcd-1 PV, removing the pv-protection finalizer in the process, also deleted the etcd-1 pod then (not sure if that would've been necessary). Eventually the PV deletion went through, the kubermatic_volume_watcher_controller kicked in, recreated the pvc and pv and restarted all the etcd pods, end result was a working 3-node cluster with all pods restarted and the etcd-1 PVC and PV recreated. +1

Glad to hear that! You don't actually need to delete anything except the PV (to test). After the PV is removed the recovery process should trigger automatically and handle everything.

moelsayed committed:
fix tests

moelsayed commented:
/test pre-kubermatic-e2e-aws-flatcar-1.18-ce

multi-io commented:
Since you removed the [WIP] label -- AFAICS the the code currently would try to fix non-etcd statefulsets as well -- and probably not find the PVC's pod and error-return [here](https://github.com/kubermatic/kubermatic/pull/5746/files#diff-5c2013a72c412b88448723c786ec3326R112). Also, worker-name and cluster paused checks are still missing.

moelsayed commented:
/hold

moelsayed committed:
review comments

moelsayed commented:
/test pre-kubermatic-e2e-aws-coreos-1.18

moelsayed commented:
/test pre-kubermatic-e2e-aws-coreos-1.18

moelsayed commented:
/test pre-kubermatic-e2e-aws-coreos-1.18

moelsayed commented:
/test pre-kubermatic-e2e-aws-coreos-1.16

moelsayed commented:
/test pre-kubermatic-e2e-aws-coreos-1.18

moelsayed commented:
/retest

moelsayed commented:
/test pre-kubermatic-e2e-aws-coreos-1.18

multi-io commented:
/lgtm

xrstf commented on pkg/controller/seed-controller-manager/pvwatcher/pvwatcher_controller.go:
```diff
@@ -0,0 +1,174 @@
+	if err := c.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, &handler.EnqueueRequestForObject{},
+		LostClaimPredicates,
+		predicateutils.ByLabel(resources.AppLabelKey, resources.EtcdStatefulSetName)); err != nil {
+		return fmt.Errorf("failed to create watch fordocker-build PersistentVolumeClaidocker-buildms: %v", err)
```
I suspect some copy-paste mistakes happened in this line?

xrstf commented on pkg/controller/seed-controller-manager/pvwatcher/pvwatcher_controller.go:
```diff
@@ -0,0 +1,174 @@
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		log.Errorf("failed to get pvc: %v", err)
```
Why is this logged explicitly, when the return value of `reconcile()` will be logged in `Reconcile()` anyway?

xrstf commented on pkg/controller/seed-controller-manager/pvwatcher/pvwatcher_controller.go:
```diff
@@ -0,0 +1,174 @@
+		return reconcile.Result{}, err
+	}
+
+	// When the pod is deleted and recreared it, it will be stuck in PodPending phase
```
_recreaTed_

xrstf reviewed (changes requested):
Minor things

moelsayed committed:
review comments

xrstf reviewed (approved):
/approve

moelsayed commented:
/retest

moelsayed commented:
/hold cancel
